src/pybind/mgr/dashboard/controllers/cephfs.py

```diff
@@ -404,34 +416,25 @@ def rm_dir(self, fs_id, path):
         cfs = self._cephfs_instance(fs_id)
         cfs.rm_dir(path)
 
-    @RESTController.Resource('POST')
-    def mk_snapshot(self, fs_id, path, name=None):
-        """
-        Create a snapshot.
-        :param fs_id: The filesystem identifier.
-        :param path: The path of the directory.
-        :param name: The name of the snapshot. If not specified,
-            a name using the current time in RFC3339 UTC format
-            will be generated.
-        :return: The name of the snapshot.
-        :rtype: str
-        """
-        cfs = self._cephfs_instance(fs_id)
-        return cfs.mk_snapshot(path, name)
 
-    @RESTController.Resource('POST')
-    def rm_snapshot(self, fs_id, path, name):
+@ApiController('/cephfs/quota', Scope.CEPHFS)
+class CephFSQuotaController(RESTController):
+    RESOURCE_ID = 'fs_quota'
+
+    @RESTController.Resource('PUT')
+    def quota(self, fs_id, path, max_bytes=None, max_files=None):
         """
-        Remove a snapshot.
+        Set the quotas of the specified path.
         :param fs_id: The filesystem identifier.
-        :param path: The path of the directory.
-        :param name: The name of the snapshot.
+        :param path: The path of the directory/file.
+        :param max_bytes: The byte limit.
+        :param max_files: The file limit.
         """
         cfs = self._cephfs_instance(fs_id)
-        cfs.rm_snapshot(path, name)
+        return cfs.set_quotas(path, max_bytes, max_files)
 
-    @RESTController.Resource('GET')
-    def get_quotas(self, fs_id, path):
+    @RESTController.Resource('GET')  # noqa: F811
+    def quota(self, fs_id, path):  # noqa: E0102
```

Comment:
Unless some [multimethod pattern](https://www.artima.com/weblogs/viewpost.jsp?thread=101605) is implemented this will cause errors, because the 2nd definition overrides the first one. Keep in mind that a decorator is basically the same as doing this:
```py
def quota():
  # PUT one
  ...
quota = RESTController.Resource('PUT')(quota)

def quota():
  # GET one
  ...
quota = RESTController.Resource('GET')(quota)
```

So only the last one is preserved. This is worked around with multimethods because the decorator used for creating a multimethod does not return a regular function, but a callable (that's it, an object that behaves as a function, but which keeps an internal state, basically a registry of all the duplicate methods).

Reply:
Should I use multimethods or just name the functions differently?

Reply:
Not in this case (a different situation is API versioning). Here the thing is that a `Quota` is a sub-resource of every CephFS `Path` sub-resource (which, is in turn, a sub-resource of the CephFS `Filesystem` root resource).

The main issue with the `controller/cephfs.py` is that it lacks proper OOP modelling: there's a single root resource (`CephFS`) and every other logical resource is handled as a sub-method of it.

A proper RESTful hierarchy would be like this: `/api/cephfs/{filesystem_id}/path/{path_id_URI_encoded}/quota/`. So this quota should be better represented by a stand-alone class rather than a method of another class:
```py
@ApiController('/cephfs
class Quota():
  def get():
    ...

  def set():
    ...

  def delete():
    ...
```

You can better check `controller/rbd.py` as an example of this OOP modelling (you have separate resources for `RBDSnapshot`, `RBDTrash`, `RBDNamespace`, ...).